### ISSUE TITLE
Setting idle time after executing on entry code

### ISSUE THREAD

**Issue by trulede** [AlexandreDecan/sismic#70]
With reference to the following code block:
   ```
 def execute_on_entry(self, state: StateMixin) -> List[Event]:
        """
        Execute the on entry action for given state.
        This method is called for every state that is entered, even those with no *on_entry*.

        :param state: the considered state
        :return: a list of sent events
        """
        self._entry_time[state.name] = self._interpreter.time
        self._idle_time[state.name] = self._interpreter.time

        return self._execute_code(getattr(state, 'on_entry', None))
```
when the on_entry code takes some time both `After(x)` and `Idle(x)` in guard conditions operate from the same reference point. If the _idle_time were set after the on_entry code was executed then the Idle(x) time would provide a more useful measure of how long the state has been waiting since the on_entry code was executed - currently its not possible to achieve that.

For instance:
  ```
        self._entry_time[state.name] = self._interpreter.time
        foo = self._execute_code(getattr(state, 'on_entry', None))
        self._idle_time[state.name] = self._interpreter.time
        return foo
```

Possibly the same change is needed in `execute_action()` where again the idle time is set before executing the action code, which if that action code takes some time, the idle time is perhaps not correct (since it includes the time for executing the action).

**Comment by AlexandreDecan** [comment]
*idle* represents the time since the latest transition (external or internal one) was processed. As such, it is expected that *idle time* and *entry time* are equivalent unless an internal transition was processed (in that case, *idle time* is reset while *entry time* does not change). 

While I understand that "some time" elapses between the moment when the value is set and the end of the execution of the *on_entry* action, this shouldn't be visible (nor used) as the "current time" of the interpreter cannot be changed between the execution of the two lines (micro steps are expected to be atomic steps according to the semantics). 

If a long running operation has to be processed by a state, I suggest to rely on an internal transition instead of the *on_entry* action, e.g., 
```yaml
on entry: done = False
transitions:
  - guard: not done
    action: long_computation()
```

**Comment by AlexandreDecan** [comment]
I'll do some changes w.r.t. when *entry time* is computed based on your observations. In Sismic, a state is entered (i.e. added to the active configuration) **after** its *on entry* action is executed. To be coherent when long running actions are processed, the *entry time* of a state should be set **after** its *on entry* action. 

I'll bring this change (and possibly some other related ones) with the next release.

**Comment by trulede** [comment]
Hi,

yeah, we have a call to another system, that call does polling on a variable in Redis, normally its fast at under 5 mS, but sometimes it takes 100+ mS (or even longer if its waiting for something to complete). Its a distributed system, so there is a bit of slack needed :)

But anyway, thats a long story ... what we did observe was that occasionally the guard conditions evaluated "too soon", and after trying both after & idle (and noticing no difference in behaviour) that is how the question arose, when exactly should these two times be set? Before or after the code is executed?

Thanks for taking a look.

**Comment by AlexandreDecan** [comment]
Hi, 

There's only a subtle difference between the *entry time* and *idle time*. The first one is used for *after(x)* that holds if *x* seconds have elapsed since the state was entered, while the second one is used for *idle(x)* that holds if *x* seconds have elapsed since the last transition for that state was processed. There's only one situation where *after(x)* and *idle(x)* do not agree: when an internal transition is processed for the state. In that case, as the state is not left, *after(x)* still holds, while *idle(x)* is reset. 

I'll change the moment when they are set/reset: 
 - *entry time* and *idle time* will be set to 0 when the state is entered **after*** the entry action is performed (currently, they are set **before**). The reason is that a state is considered to be active only when its entry action is done. 
 - *idle time* is reset to 0 when a transition is processed, but it will be reset **after** the action of that transition is performed (currently, it is set **before**). As such, both behaviours will be coherent. 

I expect to release those changes (with transition priorities and a new module containing primitives to support basic unit testing) tomorrow or Wednesday. I'll keep you informed!

**Comment by trulede** [comment]
Hi

did it ever occur to set the time on the Interpreter internally, rather than from the controlling script? That is to say, rework the time setter and have the Interpreter run on UTC at each MacroStep:

`interpreter.time = time.time() 
`
```
self.time = time.time()
macro_step = MacroStep(time=self.time, steps=executed_steps))?
```

For me it makes a bit more sense to have the Interpreter run on UTC, and then let the controlling script present that time how it wants - i.e. start offset (as in in the examples) or conversion to local time for printing. I actually use your `run_in_background` call, with a modification `max_steps = 1` (as optional parameter), to similar effect.

**Comment by AlexandreDecan** [comment]
> did it ever occur to set the time on the Interpreter internally, rather 
> than from the controlling script? That is to say, rework the time setter 
> and have the Interpreter run on UTC at each MacroStep:
> For me it makes a bit more sense to have the Interpreter run on UTC, and 
> then let the controlling script present that time how it wants - i.e. 
> start offset (as in in the examples) or conversion to local time for 
> printing. I actually use your |run_in_background| call, with a 
> modification |max_steps = 1| (as optional parameter), to similar effect.

When we implemented Sismic, we didn't have such a need, mainly because for research purpose, we wanted to have total control on the timer hence the choice of a discrete simulated clock. Having simulated time allows one to synchronize that simulated time with real time (as in some of our examples, and as in your current use case). The converse however is not true: we cannot easily simulate time based on a real clock.

**Comment by AlexandreDecan** [comment]
(Sorry, this issue was automatically closed by GitHub because of the commit message. I suggest to open a new issue about the timer, so we can specifically discuss this. If you have other questions/needs, please open issues for them, so we can also discuss about them).

**Comment by AlexandreDecan** [comment]
Based on your comment about `interpreter.time`, I'm currently working on a new way to deal both with simulated and real time. I keep you informed, I'll probably push a PR this afternoon or tomorrow to have your feedback, as I hope to bundle those changes with the next 1.3.0 release.

**Comment by AlexandreDecan** [comment]
The PR is now available: #71

@trulede could you please have a look at that PR? Thank you!
